### PR TITLE
Flow EVM versus Ethereum

### LINKED ISSUE #821: Differences between Flow EVM and Ethereum
We have covered many topics from Flow perspective (e.g. [the filp](https://github.com/onflow/flips/blob/main/protocol/20231116-evm-support.md) ), but we need some docs on differences between FLOW EVM and Ethereum, we have received many questions in the past and I want to aggregate them into a single document mostly targeting EVM devs. 

Subsections will be added soon.

### PR DESCRIPTION
**Pull request by ramtinms**
closes #821

### CODE REVIEW

**Comment by ramtinms** [comment]
@franklywatson any topic/question that you have recently heard that I might be missing in this doc?

File: docs/evm/ethereum-vs-flow-evm.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+title: Differences between EVM on Flow and Ethereum
+sidebar_label: EVM on Flow vs. Ethereum
+sidebar_position: 14
+---
+
+# Differences between EVM on Flow and Ethereum
+
+## Introduction
+
+The Flow network uses [Cadence](https://cadence-lang.org/) as its main execution environment. Cadence offers a safer, more efficient, and more developer-friendly experience for building smart contracts and decentralized applications. However, supporting EVM could make it easier for existing applications and tools already deployed for the EVM ecosystem to onboard to the Flow network ([more](https://developers.flow.com/evm/about) on this topic).
```
Comment (Aliserag) [inline docs/evm/ethereum-vs-flow-evm.md:11]: Lets hyperlink EVM and remove the brackets for conciseness.

File: docs/evm/ethereum-vs-flow-evm.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+EVM on Flow is designed with these major goals in mind: 
+
+- Supporting EVM equivalency: Ensure that any tools and applications deployed to or run on Ethereum can also be deployed and run on Flow.
+- Minimizing breaking changes to the Cadence ecosystem, software and tools
+- Maximum composability across environments: Allowing atomic and smooth interaction between EVM and Cadence environments.
+
+Let's look at some aspects of design to gain a better understanding:
+
+### EVM - a smart contract in Cadence
```
Comment (Aliserag) [inline docs/evm/ethereum-vs-flow-evm.md:21]: Title case for the headings to be consistent with rest of the docs

File: docs/evm/ethereum-vs-flow-evm.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+The Flow network uses [Cadence](https://cadence-lang.org/) as its main execution environment. Cadence offers a safer, more efficient, and more developer-friendly experience for building smart contracts and decentralized applications. However, supporting EVM could make it easier for existing applications and tools already deployed for the EVM ecosystem to onboard to the Flow network ([more](https://developers.flow.com/evm/about) on this topic).
+
+EVM on Flow is designed with these major goals in mind: 
+
+- Supporting EVM equivalency: Ensure that any tools and applications deployed to or run on Ethereum can also be deployed and run on Flow.
+- Minimizing breaking changes to the Cadence ecosystem, software and tools
+- Maximum composability across environments: Allowing atomic and smooth interaction between EVM and Cadence environments.
+
+Let's look at some aspects of design to gain a better understanding:
```
Comment (Aliserag) [inline docs/evm/ethereum-vs-flow-evm.md:19]: Don't think we need this sentence. The colon gives the impression that there will be a list but the format of the rest of the doc and differing header sizes makes it a bit strange.

File: docs/evm/ethereum-vs-flow-evm.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+
+### EVM - a smart contract in Cadence
+
+To satisfy the design goals and thanks to the extensibility properties of the Cadence runtime, EVM on Flow is designed as a higher-level environment incorporated as a smart contract deployed to Cadence. This smart contract is not owned by anyone and has its own storage space, allows Cadence to query, and is updated through EVM transactions. EVM transactions can be wrapped inside Cadence transactions and passed to the EVM contract for execution. The artifacts of EVM transaction execution (e.g. receipts and logs) are emitted as special Cadence events (TransactionExecuted, BlockExecuted) and available to the upstream process (Flow transaction) to enable atomic operations. 
+
+The EVM environment has its own concept of blocks, and a single Flow block can include several EVM blocks (a one-to-many relationship). Note that since EVM blocks are formed on-chain and Flow provides fast finality, as long as the user of these events waits for Flow block finality, it doesn’t have to worry about EVM block forks, uncle chains, and other consensus-related problems on Ethereum. 
+
+### No shared memory
```
Comment (Aliserag) [inline docs/evm/ethereum-vs-flow-evm.md:27]: Title case

File: docs/evm/ethereum-vs-flow-evm.md [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+The interaction between two environments is through atomic calls and none of the environments has access to the raw memory of the other. This maintains the security properties of each environment. Cadence can submit transactions to the EVM environment and EVM transactions can make calls to a special precompiled contract called Cadence Arch.  
+
+### No new native token
```
Comment (Aliserag) [inline docs/evm/ethereum-vs-flow-evm.md:31]: Title case

File: docs/evm/ethereum-vs-flow-evm.md [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+EVM on Flow uses the same native token as Cadence (FLOW token). No new token is minted at the genesis block of EVM and all the tokens have to be bridged over from the Cadence side into the EVM side. To facilitate this a native bridge is provided by the EVM contract. 
+
+### EVM equivalency
```
Comment (Aliserag) [inline docs/evm/ethereum-vs-flow-evm.md:35]: Title case

File: docs/evm/ethereum-vs-flow-evm.md [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+- `FlowBlockHeight() uint64` (signature: `0x53e87d66`) returns the current flow block height, this could be used instead of flow evm block heights to trigger scheduled actions given it's more predictable when a block might be formed.
+- `VerifyCOAOwnershipProof(bytes32 _hash, bytes memory _signature)(bool success)` returns true if the proof is valid. An ownership proof verifies that a Flow wallet controls a COA account (see the next section for more details on COA).
+- `getRandomSource(uint64) bytes32` returns a secure on-chain random source. This can be used for the creation of PRGs (learn more about [secure random on Flow here](https://developers.flow.com/build/advanced-concepts/randomness)).
+- `revertibleRandom() uint64` returns a pseudo-random value that is produced by the Flow PRG and uses a secure on-chain random source with salt as the seed to the generator. Using this random value is safe, but it allows for a transaction to revert in case of an unfavourable outcome, so it must be used carefully and it's best to [follow safety guidelines](https://developers.flow.com/build/advanced-concepts/randomness).
+
+## Special addresses
```
Comment (Aliserag) [inline docs/evm/ethereum-vs-flow-evm.md:115]: Title case

File: docs/evm/ethereum-vs-flow-evm.md [hunk 8]
```diff
@@ -0,0 +1,130 @@
+---
+title: Differences between EVM on Flow and Ethereum
+sidebar_label: EVM on Flow vs. Ethereum
+sidebar_position: 14
+---
+
+# Differences between EVM on Flow and Ethereum
```
Comment (Aliserag) [inline docs/evm/ethereum-vs-flow-evm.md:7]: Title case

File: docs/evm/ethereum-vs-flow-evm.md [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+
+The EVM environment has its own concept of blocks, and a single Flow block can include several EVM blocks (a one-to-many relationship). Note that since EVM blocks are formed on-chain and Flow provides fast finality, as long as the user of these events waits for Flow block finality, it doesn’t have to worry about EVM block forks, uncle chains, and other consensus-related problems on Ethereum. 
+
+### No shared memory
+
+The interaction between two environments is through atomic calls and none of the environments has access to the raw memory of the other. This maintains the security properties of each environment. Cadence can submit transactions to the EVM environment and EVM transactions can make calls to a special precompiled contract called Cadence Arch.
```
Comment (Aliserag) [inline docs/evm/ethereum-vs-flow-evm.md:29]: Can we link to Cadence Arch

File: docs/evm/ethereum-vs-flow-evm.md [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+Yet a small set of differences between EVM on Flow and Ethereum might be seen (mostly of the nature of extension) for two reasons: 
+
+- A set of extensions has been added to ensure seamless and easy interaction between the two environments.
+- EVM on Flow is secured by the Flow network and benefits from its nice properties. For example, Flow supports fast block production and finalization, making issues like handling uncle chains irrelevant.
+
+### Gateways
+
+As mentioned earlier EVM on Flow doesn’t have its own consensus and runs on top of the Flow network. There is no new node type for the EVM on Flow. Anyone who wants to run the network can stake tokens and participate as Flow nodes.
+
+Gateway is a new software package provided by the Flow Foundation. It communicates with the Flow access nodes and can be run by anyone (unstaked). It serves two purposes:
```
Comment (Aliserag) [inline docs/evm/ethereum-vs-flow-evm.md:48]: Lets link to resources on the EVM Gateway when we introduce it

Comment (Aliserag) [inline docs/evm/ethereum-vs-flow-evm.md:48]: Lets reformat this a bit as the two purposes we highlight have headers that are the same size as the following 'Censorship resistance and MEV protection' section making it visually harder to follow

File: docs/evm/ethereum-vs-flow-evm.md [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+
+Note that the safety of transaction execution is not dependent on the gateway; they only relay the transaction. The safety measures of the EVM environment (e.g., Nonce, etc.) ensure that each transaction is executed at most once. Since gateways are submitting Flow transactions, they have to pay the related transaction fees. Part of these fees is associated with the computation fees of the EVM transaction.
+
+To facilitate the repayment of fees, the EVM run transaction function accepts a coinbase address, collects the gas fees from the transaction, and pays it to the address provided by the gateway node. Essentially, the transaction wrapper behaves similarly to a miner, receives the gas usage fees on an EVM address, and pays for the transaction fees. The gas price per unit of gas creates a marketplace for these 3rd parties to compete over transactions.
```
Comment (Aliserag) [inline docs/evm/ethereum-vs-flow-evm.md:60]: Lets italize or code block 'run transaction' to make it clear that its a name.

File: docs/evm/ethereum-vs-flow-evm.md [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+
+**Censorship resistance and MEV protection**
+
+Since EVM on Flow runs on the Flow network, it benefits from Flow’s protections against censorship and MEV attacks. The Flow network natively provides censorship & MEV resistance which is achieved by designating specific validators for building transaction bundles that are separated from the validators proposing blocks (proposer-builder separation). You can read more about it on the Flow’s protocol white paper. 
+For extra protection on the EVM side, the gateway software is designed to be fully configurable and as lightweight as possible. This enables anyone with an account on Flow (e.g., any application) to run this software.
+
+**EIP-1559**
```
Comment (Aliserag) [inline docs/evm/ethereum-vs-flow-evm.md:67]: Would add the full name: Fee Market Change (EIP-1559)

File: docs/evm/ethereum-vs-flow-evm.md [hunk 13]
```diff
@@ -0,0 +1,130 @@
+---
+title: Differences between EVM on Flow and Ethereum
```
Comment (Aliserag) [inline docs/evm/ethereum-vs-flow-evm.md:2]: Given that we already have docs on the difference between EVM that still need to be refactored (https://docs-git-ramtin-821-onflow.vercel.app/build/differences-vs-evm) and the content I would recommend renaming this doc to 'How it Works' or 'How EVM on Flow Works'.

File: docs/evm/ethereum-vs-flow-evm.md [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+sidebar_label: EVM on Flow vs. Ethereum
+sidebar_position: 14
```
Comment (Aliserag) [inline docs/evm/ethereum-vs-flow-evm.md:4]: Lets put this under 'Why EVM on Flow' and change the position to 2. It is a more logical grouping given content.

File: docs/evm/ethereum-vs-flow-evm.md (hunk 1 above)
Comment (ramtinms) [reply docs/evm/ethereum-vs-flow-evm.md:11]: :-? not sure if I follow isn't it the bracket and url the way we construct hyperlinks in markdown?

File: docs/evm/ethereum-vs-flow-evm.md (hunk 3 above)
Comment (ramtinms) [reply docs/evm/ethereum-vs-flow-evm.md]: removed

File: docs/evm/ethereum-vs-flow-evm.md (hunk 2 above)
Comment (ramtinms) [reply docs/evm/ethereum-vs-flow-evm.md]: updated

File: docs/evm/ethereum-vs-flow-evm.md (hunk 5 above)
Comment (ramtinms) [reply docs/evm/ethereum-vs-flow-evm.md]: updated

File: docs/evm/ethereum-vs-flow-evm.md (hunk 11 above)
Comment (ramtinms) [reply docs/evm/ethereum-vs-flow-evm.md]: updated

File: docs/evm/ethereum-vs-flow-evm.md (hunk 4 above)
Comment (ramtinms) [reply docs/evm/ethereum-vs-flow-evm.md]: updated

File: docs/evm/ethereum-vs-flow-evm.md (hunk 6 above)
Comment (ramtinms) [reply docs/evm/ethereum-vs-flow-evm.md]: updated

File: docs/evm/ethereum-vs-flow-evm.md (hunk 7 above)
Comment (ramtinms) [reply docs/evm/ethereum-vs-flow-evm.md]: updated

File: docs/evm/ethereum-vs-flow-evm.md (hunk 8 above)
Comment (ramtinms) [reply docs/evm/ethereum-vs-flow-evm.md:7]: updated

File: docs/evm/ethereum-vs-flow-evm.md (hunk 10 above)
Comment (ramtinms) [reply docs/evm/ethereum-vs-flow-evm.md:46]: updated

File: docs/evm/ethereum-vs-flow-evm.md (hunk 14 above)
Comment (ramtinms) [reply docs/evm/ethereum-vs-flow-evm.md]: updated

File: docs/evm/ethereum-vs-flow-evm.md (hunk 10 above)
Comment (ramtinms) [reply docs/evm/ethereum-vs-flow-evm.md:46]: When this becomes available we can update this doc.

File: docs/evm/ethereum-vs-flow-evm.md (hunk 13 above)
Comment (ramtinms) [reply docs/evm/ethereum-vs-flow-evm.md]: updated

File: docs/evm/ethereum-vs-flow-evm.md (hunk 12 above)
Comment (ramtinms) [reply docs/evm/ethereum-vs-flow-evm.md]: updated

**Comment by ramtinms** [comment]
@Aliserag ready for another round of review.

**Review by SeanRobb — CHANGES_REQUESTED** [review]
Can we update the intro paragraph slightly. 

We should promote Cadence while also promoting EVM on Flow.

File: docs/evm/how-it-works.md [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+title: Differences How EVM on Flow Works
+sidebar_label: How it Works
+sidebar_position: 2
+---
+
+# Differences between EVM on Flow and Ethereum
+
+## Introduction
+
+The Flow network uses [Cadence](https://cadence-lang.org/) as its main execution environment. Cadence offers a safer, more efficient, and more developer-friendly experience for building smart contracts and decentralized applications. However, supporting EVM could make it easier for existing applications and tools already deployed for the EVM ecosystem to onboard to the Flow network ([more](https://developers.flow.com/evm/about) on this topic).
```
[the remark this answers is not in the document]
Comment (ramtinms) [reply docs/evm/how-it-works.md]: updated at [ae56996](https://github.com/onflow/docs/pull/827/commits/ae56996b0bd1d295cb0a35be00d1bc659a04e508)

File: docs/evm/how-it-works.md [hunk 16]
```diff
@@ -0,0 +1,129 @@
+---
+title: Differences How EVM on Flow Works
```
Comment (Aliserag) [inline docs/evm/how-it-works.md:2]: Would update this as well to: How EVM on Flow Works

File: docs/evm/how-it-works.md [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+sidebar_label: How it Works
+sidebar_position: 2
+---
+
+# Differences between EVM on Flow and Ethereum
```
Comment (Aliserag) [inline docs/evm/how-it-works.md:7]: Would update this as well to: How EVM on Flow Works

File: docs/evm/how-it-works.md [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+
+## Introduction
+
+The Flow network uses [Cadence](https://cadence-lang.org/) as its main execution environment. Cadence offers a safe, efficient, and developer-friendly experience for building smart contracts and decentralized applications. With Cadence offering full EVM support existing applications and tools already deployed for the EVM ecosystem can simply onboard to the Flow network with no code changes ([more](https://developers.flow.com/evm/about) on this topic).
```
Comment (Aliserag) [inline docs/evm/how-it-works.md:11]: Typo, would update to:

The Flow network uses [Cadence](https://cadence-lang.org/) as its main execution environment. Cadence offers a safe, efficient, and developer-friendly experience for building smart contracts and decentralized applications. Cadence can be used to extend EVM apps built in solidity by unlocking gaslesss experiences, new business models and fine-tuned access control. With Flow offering full EVM support, existing applications and tools already deployed in the EVM ecosystem can simply onboard to the network with [no code changes](https://developers.flow.com/evm/about).

File: docs/evm/how-it-works.md [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+
+To satisfy the design goals and thanks to the extensibility properties of the Cadence runtime, EVM on Flow is designed as a higher-level environment incorporated as a smart contract deployed to Cadence. This smart contract is not owned by anyone and has its own storage space, allows Cadence to query, and is updated through EVM transactions. EVM transactions can be wrapped inside Cadence transactions and passed to the EVM contract for execution. The artifacts of EVM transaction execution (e.g. receipts and logs) are emitted as special Cadence events (TransactionExecuted, BlockExecuted) and available to the upstream process (Flow transaction) to enable atomic operations. 
+
+The EVM environment has its own concept of blocks, and every Flow block includes a at most one EVM Block. The EVM block is formed at the end of Flow Block execution and includes all the transaction executed during the EVM block execution. Note that since EVM blocks are formed on-chain and Flow provides fast finality, as long as the user of these events waits for Flow block finality, it doesn’t have to worry about EVM block forks, uncle chains, and other consensus-related challenges.
```
Comment (Aliserag) [inline docs/evm/how-it-works.md:24]: typo: remove 'a' in 'a at most'

File: docs/evm/how-it-works.md [hunk 20]
```diff
@@ -0,0 +1,129 @@
...
+
+A COA is **not** controlled by a key and instead every COA account has a unique resource accessible on the Cadence side, and anyone who owns that resource and submits transactions on behalf of this address. These direct transactions have COA’s EVM address as the `tx.origin` and a new EVM transaction type (`TxType = 0xff`) is used to differentiate these transactions from other types of EVM transactions (e.g, DynamicFeeTxType (`0x02`). Currently, to make integration and tracking of these transactions byte EVM ecosystem tools, these types of transactions are encoded as legacy EVM transactions (hash computation is based on legacy tx rlp encoding).
+Controlling through a resource makes COA powerful smart contract wallets. It makes the transfer of ownership of the EVM address super easy without the need to transfer all the assets that an EVM address owns. It also allows a Cadence smart contract to take ownership of an EVM address and makes fully decentralized exchange and bridges across environments possible.
```
Comment (Aliserag) [inline docs/evm/how-it-works.md:109]: Either, 'make COAs' or 'makes a COA'

File: docs/evm/how-it-works.md [hunk 21]
```diff
@@ -0,0 +1,129 @@
...
+
+These smart contract wallets are only deployable through the Cadence environment and their address starts with the prefix `0x000000000000000000000002`. The address `0x0000000000000000000000020000000000000000` is reserved for COA factory, an address that deploys contracts for COA accounts.
+
+A COA is **not** controlled by a key and instead every COA account has a unique resource accessible on the Cadence side, and anyone who owns that resource and submits transactions on behalf of this address. These direct transactions have COA’s EVM address as the `tx.origin` and a new EVM transaction type (`TxType = 0xff`) is used to differentiate these transactions from other types of EVM transactions (e.g, DynamicFeeTxType (`0x02`). Currently, to make integration and tracking of these transactions byte EVM ecosystem tools, these types of transactions are encoded as legacy EVM transactions (hash computation is based on legacy tx rlp encoding).
```
Comment (Aliserag) [inline docs/evm/how-it-works.md:108]: Run on sentence, better:  "A COA is not controlled by a key. Instead, every COA account has a unique resource accessible on the Cadence side, and anyone who owns that resource submits transactions on behalf of this address."

File: docs/evm/how-it-works.md [hunk 22]
```diff
@@ -0,0 +1,129 @@
...
+
+To learn more about how to interact with a COA from the Cadence side, [see here](https://developers.flow.com/evm/cadence/interacting-with-coa).
+
+## Proofs
+
+**Inclusion/non-inclusion proof of execution artifacts (logs and receipts)**
+
+Similar to other EVM environments, proof can be constructed for artifacts such as receipts. As mentioned earlier, all the EVM execution artifacts are collected as part of a Cadence event. Cadence events are similar to EVM logs, and the root hash of all events (event commitment) of a block is included in Flow's block content. The Cadence event inclusion proof functionality enables constructing proofs for any artifact. For example, if one wants to construct an external proof for the inclusion of a specific EVM log or receipt, here are the steps:
+
+- Flow block validation: Anyone following the Flow blocks have can validates the Flow block headers.
```
Comment (Aliserag) [inline docs/evm/how-it-works.md:119]: Typo, should be, "Flow block validation: Anyone following the Flow blocks can validate the Flow block headers."

File: docs/evm/how-it-works.md [hunk 23]
```diff
@@ -0,0 +1,129 @@
...
+
+- EVM block validation: Since each Flow block has the root hash of Cadence events emitted during block execution. It can construct and verify the inclusion of the specific Event. In this case, every time an EVM block is executed an `evm.BlockExecuted` event is emitted that contains the full EVM block information.
+
+- Receipt inclusion** : Each EVM block includes the root hash for the receipts generated during block execution. Similar to other EVM chains, a proof can be constructed to prove inclusion of a log or receipt.
```
Comment (Aliserag) [inline docs/evm/how-it-works.md:123]: ** seems like a typo

Comment (ramtinms) [reply docs/evm/how-it-works.md]: fixed

File: docs/evm/how-it-works.md (hunk 22 above)
Comment (ramtinms) [reply docs/evm/how-it-works.md]: fixed

File: docs/evm/how-it-works.md (hunk 21 above)
Comment (ramtinms) [reply docs/evm/how-it-works.md]: updated

File: docs/evm/how-it-works.md (hunk 20 above)
Comment (ramtinms) [reply docs/evm/how-it-works.md]: fixed

File: docs/evm/how-it-works.md (hunk 19 above)
Comment (ramtinms) [reply docs/evm/how-it-works.md]: fixed

File: docs/evm/how-it-works.md (hunk 18 above)
Comment (ramtinms) [reply docs/evm/how-it-works.md]: updated

File: docs/evm/how-it-works.md (hunk 17 above)
Comment (ramtinms) [reply docs/evm/how-it-works.md]: fixed

File: docs/evm/how-it-works.md (hunk 16 above)
Comment (ramtinms) [reply docs/evm/how-it-works.md]: fixed

**Review by gregsantos — APPROVED** [review]

**Pull request merged**